---
title: "Differential OTUs"
author: "Hui/Jonathan"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: word_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, include = FALSE)
setwd("B:/")
source("fcns/config.r")
source("fcns/select_top.r")
source("fcns/fcns_limma/edger_contrasts.r")
setwd("external/altindis/celiac_disease_16sn/edger")
library(plyr)
library(reshape2)
```

## Purpose
Differential OTUs by using edger [1]
Results are in folder [edger](../edger).

## Data
OTU table [otu_table.csv](../phyloseq/otu_table.csv), 
Taxonomy table [taxonomy_table.csv](../phyloseq/taxonomy_table.csv), 
Metadata table [metadata_table.csv](../phyloseq/metadata_table.csv)

```{r parse}
counts <- data.matrix(read.csv("../phyloseq/otu_table.csv", row.names = 1,  skipNul = TRUE))
pheno <- read.csv("../phyloseq/metadata_table.csv", row.names = 1,  skipNul = TRUE)
annot <- read.csv("../phyloseq/taxonomy_table.csv", row.names = 1,  skipNul = TRUE)
rownames(counts) <- rownames(annot) <- paste0("OTU", 1:nrow(counts))


annot$Taxa <- apply(annot, 1, function(v) paste(v[!is.na(v)], collapse = "_"))
pheno$Disease.status <- factor(pheno$Disease.status, levels = c("Control", "Celiac"))
pheno$Group <- factor(pheno$Group, levels = unique(pheno$Group))
```

We combine the counts of each two technical replicates.   

## Filtering
```{r filt}
dge <- DGEList(counts = counts)
dge <- dge[rowSums(counts >= 10) > 5, ]
dim(dge)

# donot using  calcNormFactors()
```

To filter out low abundant OTUs, we keep OTUs that have read counts at least 10 in 5 samples. There are 611 OTUs after filtering. 

## Test for differential OUTs
```{r edger}
contr.v <- c(Celiac_vs_Control_in_Presort_Age1 = "Presort_Celiac_1 - Presort_Control_1",
             Celiac_vs_Control_in_IGneg_Age1 = "IGneg_Celiac_1 - IGneg_Control_1",
             Celiac_vs_Control_in_IGpos_Age1 = "IGpos_Celiac_1 - IGpos_Control_1", 
             IGpos_vs_IGneg_in_Control_Age1 = "IGpos_Control_1 - IGneg_Control_1",  
             IGpos_vs_IGneg_in_Celiac_Age1 = "IGpos_Celiac_1 - IGneg_Celiac_1",
             
             Celiac_vs_Control_in_Presort_Age2.5 = "Presort_Celiac_2.5 - Presort_Control_2.5",
             Celiac_vs_Control_in_IGneg_Age2.5 = "IGneg_Celiac_2.5 - IGneg_Control_2.5",
             Celiac_vs_Control_in_IGpos_Age2.5 = "IGpos_Celiac_2.5 - IGpos_Control_2.5", 
             IGpos_vs_IGneg_in_Control_Age2.5 = "IGpos_Control_2.5 - IGneg_Control_2.5",  
             IGpos_vs_IGneg_in_Celiac_Age2.5 = "IGpos_Celiac_2.5 - IGneg_Celiac_2.5",
             
             Celiac_vs_Control_in_Presort_Age5 = "Presort_Celiac_5 - Presort_Control_5",
             Celiac_vs_Control_in_IGneg_Age5 = "IGneg_Celiac_5 - IGneg_Control_5",
             Celiac_vs_Control_in_IGpos_Age5 = "IGpos_Celiac_5 - IGpos_Control_5", 
             IGpos_vs_IGneg_in_Control_Age5 = "IGpos_Control_5 - IGneg_Control_5",  
             IGpos_vs_IGneg_in_Celiac_Age5 = "IGpos_Celiac_5 - IGneg_Celiac_5")

pdf("edger_qc_metrics.pdf")
res <- edger.contrasts(dge, grp = pheno$Group, contrasts.v = contr.v, plot = TRUE)
dev.off()
mtt <- res$mtt; logcpm <- res$logcpm
rm(res)
mtt.df <- data.frame(signif(mtt, 3), annot[rownames(mtt), ])

signif_hist(mtt, name = "signif_hist", pi0 = TRUE)
write.csv(mtt.df, "otu_stats.csv", na = "")
```

To discover the differential OTUs, we use edgeR, an R package for differential expression analysis of digital gene expression data [1]. We perform empirical Bayes quasi-likelihood F-tests for the following comparisons: between any 2 disease status at different age and sorting, or between IGpos and IGneg in different age and disease status.    
The histograms of significance are at [signif_hist.pdf](./signif_hist.pdf). If no OTUs were associated with the phenotype, we would expect the p-value histogram to be flat and all FDRs to be near one. The more associated OTUs there are, the more enrichment there is at low p-values, the lower will be the FDRs.  We also estimate the proportion of the true null hypothesis (i.e. non-significant OTUs) [2].    
OTU statistics tables for all OTUs [otu_stats.csv](./otu_stats.csv). The table contains the average logCPM of each group, p-values, FDR, log fold-change, fold-change, and taxonomy information.      

## Plots
```{r plots}
## boxplot
topotu1 <- select.top(mtt, contrasts.v = contr.v[grep("Presort", names(contr.v))])
pdf("top_otus_boxplot_presort.pdf", 9, 3)
for (otu in topotu1){
    dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort == "Presort"], pheno[pheno$Sort == "Presort", ])
     
    ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
    ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none")
    ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(~ Age)
    plot(ggp)
}
dev.off()

topotu2 <- select.top(mtt, contrasts.v = contr.v[-grep("Presort", names(contr.v))])
pdf("top_otus_boxplot_postsort.pdf", 9, 5)
for (otu in topotu2){
    dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort != "Presort"], pheno[pheno$Sort != "Presort", ])
     
    ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
    ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none")
    ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(Sort~ Age)
    plot(ggp)
}
dev.off()

## boxplot for all
pdf("all_otus_boxplot_presort.pdf", 9, 3)
for (otu in rownames(logcpm)){
    dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort == "Presort"], pheno[pheno$Sort == "Presort", ])
     
    ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
    ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none")
    ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(~ Age)
    plot(ggp)
}
dev.off()

pdf("all_otus_boxplot_postsort.pdf", 9, 5)
for (otu in rownames(logcpm)){
    dat2p <- data.frame(logCPM = logcpm[otu, pheno$Sort != "Presort"], pheno[pheno$Sort != "Presort", ])
     
    ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
    ggp <- ggp + ggtitle(paste(otu, ",", annot[otu, "Taxa"])) + theme(legend.position = "none")
    ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(Sort~ Age)
    plot(ggp)
}
dev.off()

## Clostridium XlVa
otus <- readLines("Clostridium_XlVa.txt")
samp <- rownames(pheno)[pheno$Age == "5year"]
dat2p <- cbind(t(logcpm[otus, samp]), pheno[samp, c("Disease.status","Sort")])
dat2p <- melt(dat2p, id.vars =  c("Disease.status","Sort"),  variable.name = "OTU",value.name = "logCPM")

pdf("Clostridium_XlVa_boxplot.pdf", 6, 12)
ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
ggp <- ggp + ggtitle("Age: 5 years, Clostridium XlVa") + xlab("") + theme(legend.position = "none")
ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status)) + facet_grid(OTU~Sort)
plot(ggp)
dev.off()

####pool 6
otus <- readLines("Clostridium_XlVa.txt")
samp <- rownames(pheno)[pheno$Age == "5year" & pheno$Sort == "Presort"]
dat2p <- cbind(logCPM = colMeans(logcpm[otus, samp]), pheno[samp, "Disease.status", drop = FALSE])
pval <- signif(t.test(logCPM~Disease.status, data = dat2p)$p.value, 3)

pdf("Clostridium_XlVa_boxplot2_presort.pdf", 3.5, 3.5)
ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
ggp <- ggp + ggtitle(paste("Age: 5 years, Presort, \n6 Clostridium XlVa, pval =", pval)) + xlab("") + theme(legend.position = "none")
ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status))
plot(ggp)
dev.off()

###pool all
otus <- rownames(mtt.df)[which(mtt.df$Genus == "Clostridium_XlVa")]
samp <- rownames(pheno)[pheno$Age == "5year" & pheno$Sort == "Presort"]
dat2p <- cbind(logCPM = colMeans(logcpm[otus, samp]), pheno[samp, "Disease.status", drop = FALSE])
pval <- signif(t.test(logCPM~Disease.status, data = dat2p)$p.value, 3)

pdf("Clostridium_XlVa_boxplot3_presort.pdf", 3.5, 3.5)
ggp <- ggplot(data = dat2p, mapping = aes(x = Disease.status, y = logCPM)) + theme_bw()
ggp <- ggp + ggtitle(paste("Age: 5 years, Presort, \nAll Clostridium XlVa, pval =", pval)) + xlab("") + theme(legend.position = "none")
ggp <- ggp + geom_boxplot(mapping = aes(fill = Disease.status))
plot(ggp)
dev.off()


## heatmaps
pheno.df1 <- pheno[pheno$Sort == "Presort", c("Disease.status", "Age")]
gaps.col1 <- which(diff(as.numeric(pheno.df1$Age), lag=1) != 0)
ezheat(logcpm[topotu1, pheno$Sort == "Presort"], pheno.df = pheno.df1,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu1, "Taxa"], labcols = "",
       gaps_col = gaps.col1, main = "logCPM", name = "top_otus_heat_presort", height = 8, width = 16, clip = 2)


pheno.df2 <- pheno[pheno$Sort != "Presort", c("Disease.status", "Age", "Sort")]
gaps.col2 <- which(diff(as.numeric(pheno.df2$Age), lag=1) != 0)
ezheat(logcpm[topotu2, pheno$Sort != "Presort"], pheno.df = pheno.df2,  sc = "z", reorder_rows = TRUE, labrows = annot[topotu2, "Taxa"], labcols = "",
       gaps_col = gaps.col2, main = "logCPM", name = "top_otus_heat_postsort", height = 8, width = 25, clip = 2)
```

Equal numbers of top OTUs (based on p-values) are selected from each comparison. The boxplots for top OUTs are at [top_otus_boxplot_presort.pdf](./top_otus_boxplot_presort.pdf), and [top_otus_boxplot_postsort.pdf](./top_otus_boxplot_postsort.pdf). The same sets of top OTUs are use in heatmaps [top_otus_heat_presort.pdf](./top_otus_heat_presort.pdf) and [top_otus_heat_postsort.pdf](./top_otus_heat_postsort.pdf).


```{r check}
stopifnot(rownames(counts) == rownames(annot))
stopifnot(colnames(counts) == rownames(pheno))
stopifnot(mtt.df[1, "Taxa"] == "Bacteria_Proteobacteria_Gammaproteobacteria_Enterobacteriales_Enterobacteriaceae_Citrobacter")
```

## Reference
[1] Robinson MD, McCarthy DJ, Smyth GK (2010). edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics, 26(1), 139-140.   
[2] Langaas, M, Ferkingstad, E, and Lindqvist, B (2005). Estimating the proportion of true null hypotheses, with application to DNA microarray data. Journal of the Royal Statistical Society Series B 67, 555-572.   