---
title: "Overview of 16S DNA-seq"
author: "Hui/Jonathan"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: word_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, include = FALSE)
setwd("B:/")
source("fcns/config.r")
setwd("external/altindis/celiac_disease_16sn/phyloseq")
library(phyloseq)
library(vegan)
library(microbiome)
```

## Purpose
Overview of 16S DNA-seq of gingiva samples by Phyloseq [1].  
The results are in folder [phyloseq](../phyloseq/).  

## Data
OTU table [seqtab_nochim.rds](../dada2/seqtab_nochim.rds) and taxonomy annotation [annot_add_species.rds](../dada2/annot_add_species.rds) were generated by using R package *dada2* [2]. We merge the two technical replicates for each Sample.
Metadata [CeD new label (2).csv](../data/CeD new label (2).csv) and [Celiakibarn 190906.csv](../data/Celiakibarn 190906.csv).  

```{r parse}
# meta data
md <- read.csv("../data/CeD new label (2).csv")
md_age <- read.csv("../data/Celiakibarn 190906.csv")
md <- merge(md, md_age, all = TRUE)
rm(md_age)
# md$Disease.status[!is.na(md$Age.at.diagnosis)] <- ifelse(md$Age.at.diagnosis[!is.na(md$Age.at.diagnosis)] <= 5, "A0", md$Disease.status[!is.na(md$Age.at.diagnosis)])

md <- md[order(md$New.Sample.ID), ]
md <- md[rep(1:nrow(md), each = 3),]
md$New.Sample.ID <- paste0("S", md$New.Sample.ID, c("", ".neg", ".pos"))
md$Sort <- rep(c("Presort", "IGneg", "IGpos"), time = nrow(md)/3)
md$Disease.status <- sapply(md$Disease.status, FUN = switch, A = "Celiac", B = "Control")
md$ABIS.no <- paste0("ABIS", md$ABIS.no)

md$Disease.status <- factor(md$Disease.status, levels = c("Control", "Celiac"))
md$Sorted <- md$Sort <- factor(md$Sort, levels = c("Presort", "IGneg", "IGpos"))
levels(md$Sorted) <- list(Presort = "Presort", Postsort = c("IGneg", "IGpos"))
md <- md[order(md$Sort, md$Disease.status, md$Age, md$ABIS.no), ]
md$Group <- paste(md$Sort, md$Disease.status, md$Age, sep = "_")
md$Group <- factor(md$Group, levels = unique(md$Group))
md$Age <- factor(paste0(md$Age, "year"))
md <- data.frame(md, row.names = "New.Sample.ID")

## otu
seqtab.nochim <- readRDS("../dada2/seqtab_nochim.rds")
dim(seqtab.nochim)
grp <- gsub("\\.[1-2]$", "", gsub("\\.[1-2]\\.", ".", rownames(seqtab.nochim)))
seqtab.nochim <- rowsum(seqtab.nochim, group = grp)
rm(grp)
dim(seqtab.nochim)
nms <- intersect(rownames(md), rownames(seqtab.nochim))
seqtab.nochim <- seqtab.nochim[nms, ]
md <- md[nms, ]

# taxa
taxa <- readRDS("../dada2/annot_add_species.rds")
stopifnot(colnames(seqtab.nochim) == rownames(taxa))

#make Phyloseq Object
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), sample_data(md[rownames(seqtab.nochim), ]), tax_table(taxa))
```

## Filter and transformation
```{r filt}
#Take a look at the read counts across samples
plot(sample_sums(ps))

# filt samp
# ps <-  prune_samples(sample_sums(ps) >= 5000, ps)

# normalize samp
nps <- transform_sample_counts(ps, function(x) x/sum(x))

# filt otu
keep_otu <- genefilter_sample(ps, flist = filterfun_sample(function(x) x >= 5), A = 1)
nps <- prune_taxa(keep_otu, nps)
```

We keep OTUs with at least in 5 counts in 1 sample. We then transform the read counts to proportions (relative abundance) i.e. read counts divided by total counts of each sample.  

## Plot Ordination
```{r ord}
# NMDS
set.seed(123456)
ord <- ordinate(ps, method = "NMDS", distance  = "bray", maxit = 1000, sratmax = 25,  sfgrmin = 1e-10)

pdf("nmds_plot_1.pdf", 9, 7.5)
ggp <- plot_ordination(ps, ordination = ord, type = "samples", color = "Disease.status") + theme_bw()
ggp <- ggp + geom_point(size = 2) + facet_grid(Sort ~ Age)
plot(ggp)
dev.off()

pdf("nmds_plot_2.pdf", 9, 5)
ggp <- plot_ordination(ps, ordination = ord, type = "samples", color = "Disease.status", shape = "Sort") + theme_bw()
ggp <- ggp + geom_point(size = 2) + facet_grid(Sorted ~ Age)
plot(ggp)
dev.off()

# PCA
ord <- ordinate(ps, method = "RDA", distance  = "euclidean")

pdf("pca_plot_1.pdf", 9, 7.5)
ggp <- plot_ordination(ps, ordination = ord, type = "samples", color = "Disease.status") + theme_bw()
ggp <- ggp + geom_point(size = 2) + facet_grid(Sort ~ Age) 
plot(ggp)
dev.off()

pdf("pca_plot_2.pdf", 9, 5)
ggp <- plot_ordination(ps, ordination = ord, type = "samples", color = "Disease.status", shape = "Sort") + theme_bw()
ggp <- ggp + geom_point(size = 2) + facet_grid(Sorted ~ Age)
plot(ggp)
dev.off()
```

To get an overview of sample similarity/dissimilarity, we perform Non-metric MultiDimenstional Scaling (NMDS) analysis and Principle Components Analysis (PCA). The NMDS plots are at [nmds_plot_1.pdf](./nmds_plot_1.pdf), [nmds_plot_2.pdf](./nmds_plot_2.pdf), and the PCA plots are at [pca_plot_1.pdf](./pca_plot_1.pdf), [pca_plot_2.pdf](./pca_plot_2.pdf).  

```{r}
ps.sorted <- subset_samples(ps, Sorted == "Postsort") 
ord.sorted <- ordinate(ps.sorted, method = "RDA", distance  = "euclidean")

pdf("pca_plot_3.pdf", 6.5, 3)
ggp <- plot_ordination(ps.sorted, ordination = ord.sorted, type = "samples", color = "Age", shape = "Sort") + theme_bw()
ggp <- ggp + geom_point(size = 2) + facet_grid( ~ Disease.status) 
plot(ggp)
dev.off()

pdf("pca_plot_4.pdf", 5.5, 4)
ggp <- plot_ordination(ps.sorted, ordination = ord.sorted, type = "samples", color = "Group", shape = "Sort") + theme_bw()
ggp <- ggp + geom_point(size = 2)
plot(ggp)
dev.off()
```


## Diversity estimates
```{r div}
## alpha diversities
mss <- c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher")

pdf("alpha_diversities_plot_1.pdf", 9, 7.5)
for(ms in mss) {
  ggp <- plot_richness(ps, x = "Disease.status", measures = ms) + geom_boxplot(aes(fill = Disease.status)) + theme_bw()
  ggp <- ggp + theme(legend.position = "none") + ggtitle(paste("Measure:", ms)) + facet_grid(Sort ~ Age) 
  plot(ggp)
}
dev.off()

pdf("alpha_diversities_plot_2.pdf", 9, 7.5)
for(ms in mss) {
  ggp <- plot_richness(ps, x = "Age", measures = ms) + geom_boxplot(aes(fill = Age)) + theme_bw()
  ggp <- ggp + theme(legend.position = "none") + ggtitle(paste("Measure:", ms)) + facet_grid(Sort ~ Disease.status) 
  plot(ggp)
}
dev.off()

## anova
alpha_div <- estimate_richness(ps, split = TRUE, measures = mss)
alpha_div <- data.frame(alpha_div, md[rownames(alpha_div), ])

anova_1 <- list()
for(ms in mss) {
  anova_tmp <- matrix(NA, nrow = 3, ncol = 3)
  dimnames(anova_tmp) <- list(levels(md$Sort), levels(md$Age))
  for(age in levels(md$Age)){
    for(sort in levels(md$Sort)){
      dat_tmp <- alpha_div[md$Age == age & md$Sort == sort, ]
      aov_res <- aov(as.formula(paste(ms," ~ Disease.status")), data = dat_tmp)
      anova_tmp[sort, age] <- summary(aov_res)[[1]]["Disease.status", "Pr(>F)"]
    }
  }
  anova_tmp <- data.frame(rn = rownames(anova_tmp), signif(anova_tmp, 3), check.names = FALSE)
  colnames(anova_tmp)[1] <- ""
  anova_1[[ms]] <- anova_tmp
}
writexl::write_xlsx(anova_1, "alpha_diversities_plot_1_anova_pval.xlsx")

# anova_2 <- list()
# for(ms in mss) {
#   anova_tmp <- matrix(NA, nrow = 3, ncol = 3)
#   dimnames(anova_tmp) <- list(levels(md$Sort), levels(md$Disease.status))
#   for(sort in levels(md$Sort)){
#     for(status in levels(md$Disease.status)){
#       dat_tmp <- alpha_div[md$Disease.status == status & md$Sort == sort, ]
#       aov_res <- aov(as.formula(paste(ms," ~ Age")), data = dat_tmp)
#       anova_tmp[sort, status] <- summary(aov_res)[[1]]["Age", "Pr(>F)"]
#     }
#   }
#   anova_tmp <- data.frame(rn = rownames(anova_tmp), signif(anova_tmp, 3), check.names = FALSE)
#   colnames(anova_tmp)[1] <- ""
#   anova_2[[ms]] <- anova_tmp
# }
# writexl::write_xlsx(anova_2, "alpha_diversities_plot_2_anova_pval.xlsx")

## beta diversities use nps
dis <- distance(nps, method = "bray")
mod <- betadisper(dis, group = sample_data(nps)$Group)
stopifnot(attr(dis, "Labels") == rownames(sample_data(nps)))
stopifnot(attr(mod$distances, "names") == rownames(sample_data(nps)))
dat2p <- cbind(Distance = mod$distances, sample_data(nps))

pdf("beta_diversities_plot_1.pdf", 9, 7.5)
ggp <- ggplot(data = dat2p, aes(x = Disease.status, y = Distance)) + geom_boxplot(aes(fill = Disease.status)) + theme_bw()
ggp <- ggp + theme(legend.position = "none")
ggp <- ggp + labs(y = "Distance to centroid", title = "Bray-Curtis distance") + facet_grid(Sort ~ Age)
plot(ggp)
dev.off()

pdf("beta_diversities_plot_2.pdf", 9, 7.5)
ggp <- ggplot(data = dat2p, aes(x = Age, y = Distance)) + geom_boxplot(aes(fill = Age)) + theme_bw()
ggp <- ggp + theme(legend.position = "none")
ggp <- ggp + labs(y = "Distance to centroid", title = "Bray-Curtis distance") + facet_grid(Sort ~ Disease.status)
plot(ggp)
dev.off()

anova_1 <- list()
for(dis in "Distance") {
  anova_tmp <- matrix(NA, nrow = 3, ncol = 3)
  dimnames(anova_tmp) <- list(levels(md$Sort), levels(md$Age))
  for(age in levels(md$Age)){
    for(sort in levels(md$Sort)){
      dat_tmp <- dat2p[md$Age == age & md$Sort == sort, ]
      aov_res <- aov(as.formula(paste(dis," ~ Disease.status")), data = dat_tmp)
      anova_tmp[sort, age] <- summary(aov_res)[[1]]["Disease.status", "Pr(>F)"]
    }
  }
  anova_tmp <- data.frame(rn = rownames(anova_tmp), signif(anova_tmp, 3), check.names = FALSE)
  colnames(anova_tmp)[1] <- ""
  anova_1[[dis]] <- anova_tmp
}
writexl::write_xlsx(anova_1, "beta_diversities_plot_1_anova_pval.xlsx")

# anova_2 <- list()
# for(dis in "Distance") {
#   anova_tmp <- matrix(NA, nrow = 3, ncol = 3)
#   dimnames(anova_tmp) <- list(levels(md$Sort), levels(md$Disease.status))
#   for(sort in levels(md$Sort)){
#     for(status in levels(md$Disease.status)){
#       dat_tmp <- dat2p[md$Disease.status == status & md$Sort == sort, ]
#       aov_res <- aov(as.formula(paste(dis," ~ Age")), data = dat_tmp)
#       anova_tmp[sort, status] <- summary(aov_res)[[1]]["Age", "Pr(>F)"]
#     }
#   }
#   anova_tmp <- data.frame(rn = rownames(anova_tmp), signif(anova_tmp, 3), check.names = FALSE)
#   colnames(anova_tmp)[1] <- ""
#   anova_2[[dis]] <- anova_tmp
# }
# writexl::write_xlsx(anova_2, "beta_diversities_plot_2_anova_pval.xlsx")
```

First, we estimate the alpha diversities of the different disease status at different age and sorting. We do this by using different measures, namely "Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", and "Fisher".  The plots are at [alpha_diversities_plot_1.pdf](./alpha_diversities_plot_1.pdf) and [alpha_diversities_plot_2.pdf](./alpha_diversities_plot_2.pdf). 

The corresponding ANOVA F-test p-values are at Excel sheets [alpha_diversities_plot_1_anova_pval.xlsx](./alpha_diversities_plot_1_anova_pval.xlsx) and [alpha_diversities_plot_2_anova_pval.xlsx](./alpha_diversities_plot_2_anova_pval.xlsx). For example, presorted data from kids at 1year, we test Observed alpha diversity vs disease status with an [F-test](https://en.wikipedia.org/wiki/F-test). This would be a correlation if disease status was only two groups, but originally Celiac was split in two, so there were three. An F-test tests if any of the pairwise tests are significant. Hui reported those p-values to you.  

Secondly, we estimate the beta diversities of the different disease status at different age and sorting. We use the Bray-Curtis distance. The plots are at [beta_diversities_plot_1.pdf](./beta_diversities_plot_1.pdf), and [beta_diversities_plot_2.pdf](./beta_diversities_plot_2.pdf).The corresponding ANOVA test p-values are at Excel sheets [beta_diversities_plot_1_anova_pval.xlsx](./beta_diversities_plot_1_anova_pval.xlsx) and [beta_diversities_plot_2_anova_pval.xlsx](./beta_diversities_plot_2_anova_pval.xlsx).   

## Overview of relative abundance
```{r ra}
levs <- c("Phylum", "Class", "Order", "Family", "Genus")
pdf("relative_abundance_1.pdf", 12, 7)
for(lev in levs) {
  
  nps_lev <- nps %>%
  tax_glom(taxrank = lev) %>%                                     # agglomerate at level lev
  psmelt()                                                        # Melt to long format
  
  nps_lev <- aggregate(nps_lev[, "Abundance", drop = FALSE], by = as.list(nps_lev[, c("Disease.status", "Sort", "Age", lev)]),
                       FUN = mean)                                # aggregate samples
  nps_lev <- subset(nps_lev, Abundance > 0.01)                    # filer low abundant taxa
  
  ggp <- ggplot(nps_lev, aes_string(x = "Disease.status", y = "Abundance", fill = lev)) + geom_col(width = 0.7) 
  ggp <- ggp + theme_bw() + facet_grid(Sort ~ Age) + ylab("Average relative abundance")
  plot(ggp)
}
dev.off()

pdf("relative_abundance_2.pdf", 12, 7)
for(lev in levs) {
  
  nps_lev <- nps %>%
  tax_glom(taxrank = lev) %>%                                     # agglomerate at level lev
  psmelt()                                                        # Melt to long format
  
  nps_lev <- aggregate(nps_lev[, "Abundance", drop = FALSE], by = as.list(nps_lev[, c("Disease.status", "Sort", "Age", lev)]),
                       FUN = mean)                                # aggregate samples
  nps_lev <- subset(nps_lev, Abundance > 0.01)                    # filer low abundant taxa
  
  ggp <- ggplot(nps_lev, aes_string(x = "Age", y = "Abundance", fill = lev)) + geom_col(width = 0.7) 
  ggp <- ggp + theme_bw() + facet_grid(Sort ~ Disease.status) + ylab("Average relative abundance")
  plot(ggp)
}
dev.off()


```

We agglomerate the relative abundance (proportion) for each taxonomy levels, namely "Phylum", "Class", "Order", "Family", and "Genus". After that, we calculate the average relative abundance for each disease status, and then we filter out low abundance taxa (average relative abundance < 0.01). The plots for average relative abundance are at  [relative_abundance_1.pdf](./relative_abundance_1.pdf), and [relative_abundance_2.pdf](./relative_abundance_2.pdf)  

```{r}
nps.presorted <- subset_samples(nps, Sorted == "Presort") 

levs <- c("Phylum", "Class", "Order", "Family", "Genus")
pdf("relative_abundance_3_presort.pdf", 12, 5)
for(lev in levs) {
  
  nps_lev <- nps.presorted %>%
  tax_glom(taxrank = lev) %>%                                     # agglomerate at level lev
  psmelt()                                                        # Melt to long format
  
  nps_lev <- aggregate(nps_lev[, "Abundance", drop = FALSE], by = as.list(nps_lev[, c("Disease.status", "Sort", "Age", lev)]),
                       FUN = mean)                                # aggregate samples
  nps_lev <- subset(nps_lev, Abundance > 0.01)                    # filer low abundant taxa
  
  ggp <- ggplot(nps_lev, aes_string(x = "Disease.status", y = "Abundance", fill = lev)) + geom_col(width = 0.7) 
  ggp <- ggp + theme_bw() + facet_grid( ~ Age) + ylab("Average relative abundance")
  plot(ggp)
}
dev.off()
```


## OTUs binary Venn diagrams
```{r venn}
## Disease
res <- list()
pdf("sort_disease_venn.pdf", 15,15)
par(mfrow = c(3,3))
for(sort in levels(md$Sort)){
  for(dis in levels(md$Disease.status)){
    md_tmp <- md[md$Sort == sort & md$Disease.status == dis, ]
    otu_tmp <- seqtab.nochim[rownames(md_tmp), ]
    otu_binary <- 1 * t(rowsum(otu_tmp, group = md_tmp$Age) > 0)
    nm <- paste(sort, "of", dis)
    vennDiagram(otu_binary, cex = c(1, 1, 1), circle.col = rainbow(3), counts.col = "blue", main = paste0("\n", nm))
  
    otu_binary <- otu_binary[order(-1*rowSums(otu_binary)), ]
    res_tmp <- data.frame(otu_binary, taxa[rownames(otu_binary), ], check.names = FALSE)
    res[[nm]] <- res_tmp
    }
}
dev.off()

writexl::write_xlsx(res, "sort_disease_venn_tab.xlsx")

## Age
res <- list()
pdf("sort_age_venn.pdf", 15,15)
par(mfrow = c(3,3))
for(sort in levels(md$Sort)){
  for(age in levels(md$Age)){
    md_tmp <- md[md$Sort == sort & md$Age == age, ]
    otu_tmp <- seqtab.nochim[rownames(md_tmp), ]
    otu_binary <- 1 * t(rowsum(otu_tmp, group = md_tmp$Disease.status) > 0)
    nm <- paste(sort, "of", age)
    vennDiagram(otu_binary, cex = c(1, 1, 1), circle.col = rainbow(3), counts.col = "blue", main = paste0("\n", nm))
  
    otu_binary <- otu_binary[order(-1*rowSums(otu_binary)), ]
    res_tmp <- data.frame(otu_binary, taxa[rownames(otu_binary), ], check.names = FALSE)
    res[[nm]] <- res_tmp
    }
}
dev.off()

writexl::write_xlsx(res, "sort_age_venn_tab.xlsx")

## Sort
res <- list()
pdf("disease_age_venn.pdf", 15,15)
par(mfrow = c(3,3))
for(dis in levels(md$Disease.status)){
  for(age in levels(md$Age)){
    md_tmp <- md[md$Disease.status == dis & md$Age == age & md$Sorted == "Postsort", ]
    otu_tmp <- seqtab.nochim[rownames(md_tmp), ]
    otu_binary <- 1 * t(rowsum(otu_tmp, group = md_tmp$Sort) > 0)
    nm <- paste(dis, "of", age)
    vennDiagram(otu_binary, cex = c(1, 1, 1), circle.col = rainbow(3), counts.col = "blue", main = paste0("\n", nm))
  
    otu_binary <- otu_binary[order(-1*rowSums(otu_binary)), ]
    res_tmp <- data.frame(otu_binary, taxa[rownames(otu_binary), ], check.names = FALSE)
    res[[nm]] <- res_tmp
    }
}
dev.off()

writexl::write_xlsx(res, "disease_age_venn_tab.xlsx")
```

Venn diagrams are at [sort_disease_venn.pdf](./sort_disease_venn.pdf), [sort_age_venn.pdf](./sort_age_venn.pdf), and [disease_age_venn.pdf](./disease_age_venn.pdf). The corresponding Excel sheets are at [sort_disease_venn_tab.xlsx](./sort_disease_venn_tab.xlsx), [sort_age_venn_tab.xlsx](./sort_age_venn_tab.xlsx), and [disease_age_venn_tab.xlsx](./disease_age_venn_tab.xlsx).   

```{r save}
write_phyloseq(ps,  type = "OTU")
write_phyloseq(ps,  type = "TAXONOMY")
write_phyloseq(ps,  type = "METADATA")
```

## References
[1] McMurdie PJ, Holmes S (2013). “phyloseq: An R package for reproducible interactive analysis and graphics of microbiome census data.” PLoS ONE, 8(4), e61217. http://dx.plos.org/10.1371/journal.pone.0061217.  
[2] Callahan BJ, McMurdie PJ, Rosen MJ, Han AW, Johnson AJA, Holmes SP (2016). “DADA2: High-resolution sample inference from Illumina amplicon data.” Nature Methods, 13, 581-583. doi: 10.1038/nmeth.3869.  
